controllers/storagecluster/cephcluster.go

```diff
@@ -94,6 +96,50 @@ func arbiterEnabled(sc *ocsv1.StorageCluster) bool {
 	return sc.Spec.Arbiter.Enable
 }
 
+func (r *StorageClusterReconciler) deleteJobsWithDuplicateTolerations(sc *ocsv1.StorageCluster) error {
+	// delete the osd prepare jobs which are completed and have duplicate placements
+	jobName := "rook-ceph-osd-prepare"
+	listOpts := &client.ListOptions{
+		Namespace: sc.Namespace,
+		LabelSelector: labels.SelectorFromSet(map[string]string{
+			"app": jobName,
+		}),
+	}
+
+	jobs := &batchv1.JobList{}
+	// list all the prepare jobs
+	err := r.Client.List(r.ctx, jobs, listOpts)
+	if err != nil {
+		r.Log.Error(err, "failed to list OSD prepare pod jobs")
+		return err
+	}
+
+	// filter out the completed jobs
+	for _, job := range jobs.Items {
+		if job.Status.Succeeded == 1 {
+			// filter the jobs that has duplicated tolerations
+			tolerations := job.Spec.Template.Spec.Tolerations
+
+			duplicate := make(map[corev1.Toleration]bool)
+			for _, toleration := range tolerations {
+				if duplicate[toleration] {
+					r.Log.Info("deleting completed OSD prepare jobs with duplicate tolerations", "name", klog.KRef(job.Namespace, job.Name))
+					deletePolicy := metav1.DeletePropagationForeground
+					err := r.Client.Delete(r.ctx, &job, &client.DeleteOptions{PropagationPolicy: &deletePolicy})
+					if err != nil {
+						r.Log.Error(err, "failed to delete OSD prepare pod job", "name", klog.KRef(job.Namespace, job.Name))
+						return err
+					}
+					break
+				}
+				duplicate[toleration] = true
+			}
+		}
```

Comment:
This looks good to me, but a note here. This code would have fewer indentatations and would read more simply if the loops were written using go's recommended happy-path-on-the-left semantics like this 

```go
for _, job := range jobs.Item {
    if job.Status.Succeeded == 0 {
        continue // don't delete unsuccessful jobs
    }
    
    tolerations := job.Spec.Template.Spec.Tolerations
    for _, toleration := range tolerations {
        if _, ok := duplicate[toleration]; !ok {
            duplicate[toleration] = true // record the first time seeing the toleration
        }
        
        // delete duplicate
        // etc. etc. etc.
    }
}
```

Reply:
yup that looks much cleaner
thanks